### DISCUSSION TITLE
Simulations parallelization

### DISCUSSION THREAD

**Discussion by GiM6114** [virtual-twin/tvboptim#3] [Q&A] [answered]
Hello,
I am wondering what would be the standard way to use the Axes and Spaces abstraction to run several simulations in parallel which would have different parameter values.
In the tutorial page (https://virtual-twin.github.io/tvboptim/basics/axes_and_spaces.html) there are examples for running a simple function for different parameter values in parallel using ParallelExecution, but I am struggling to link these examples to the simulations we can run as in the first page of the website (https://virtual-twin.github.io/tvboptim/) where we define a Network object, an integrator, etc.

**Comment by mapi1** [comment, marked as answer]
Thanks for asking here!

Going from the example on the first page you would start from using the `prepare` function. From there you get a pure function that JAX can do all its transformations on and a parameter container where all the values that define our simulation live. In this container we then can place axes that span a parameter space that we then can execute in parallel (see also: [RWW workflow example](https://virtual-twin.github.io/tvboptim/workflows/RWW.html):

```python
simulator, params = prepare(network, solver, t0=0.0, t1=60_000.0, dt=1.0)

# Define the parameter ranges to explore, here in a grid fashion
params.dynamics.w = GridAxis(0.0, 1.0, 10)
params.coupling.delayed.G = GridAxis(0.0, 1.0, 10)

# Create space (product creates all combinations of w and G 10x10 = 100)
grid = Space(params, mode="product")

# Parallel execution across 8 "devices"
results = ParallelExecution(simulator, grid, n_pmap=8).run()
results_df = results.to_dataframe()
```

A few things to note, creating a large space and executing it over the raw simulator can quickly fill your RAM with large time series, often it is better to define your observations of interest (e.g. FC) and only return those. If you really want to store the full time series for a large space, an approach that works is using subspaces (`subgrid = grid[n:n+100]`) in a loop and saveing results to disk in between. 
To be able to use the n_pmap option (which is originally designed for multi device setups) on a single CPU we have to use the "pmap-Trick", telling JAX / XLA in the beginning there are multiple devices by setting an environment variable:
```python
import os
os.environ['XLA_FLAGS'] = '--xla_force_host_platform_device_count=8'
``` 
This is necessary as `jax.vmap` is not much optimized on CPU. On GPU though, you would just set n_vmap and vectorization executes your simulation in parallel.

**Reply by GiM6114** [reply]
Thank you very much for your answer, I had somehow missed this RWW workflow tutorial !

**Reply by GiM6114** [reply]
Another quick question: if I wanted to parallelize simulations in terms of seed (like running in parallel with same type of noise generator but different seeds), would the "correct" implementation be to add the seed as a parameter of a noise class (since noise's seed is not a params in the default classes) ? So defining for instance AdditiveNoiseParameterizedSeed(AbstractNoise) with DEFAULT_PARAMS = Bunch(sigma=0.1, seed=0), setting the jax key in the __init__ function using this params.seed parameter ? Or would there be a more straightforward way ?

**Reply by GiM6114** [reply]
Hello again, upon closer inspection I have a surprising behavior: when running a single simulation of my custom model using solve, wall time is around 10s. However when running 2 simulations in parallel (on a 8-cores CPU, via n_pmap as you described) using ParallelExecution the total wall time is 52s! Another perhaps related issue is that when doing a loop over simulations like so :

for b_e in [5,5,5]: # no change of parameters to show it is not the issue
    network.params.dynamics.b_e = b_e
    result = solve(network, solver, t0=0.0, t1=5000.0, dt=0.1)
    
The first iteration of the loop lasts 10s and the following ones 60s! And the RAM does not seem to be filled.

I tried similar tests running the RWW model (1M steps) with delay graph & coupling, and obtain these: 
	single simulation: 1.18s
	2 parallelized simulations: 24.1s (very high again)
	4 "looped" simulations: 0.82s, 26.73s, 25.95s, 25.99s (becomes very high again)
	
And the RWW model (1M steps) without delays:
	single simulation: 747ms
	2 parallelized simulations: 11.3s (very high again)
	4 "loop" simulations: 0.61s, 0.76s, 3.22s, 3.48s (close to single simulation setup)
	
So it seems the second issue (with the loop) seems to be related to delays somehow, but the surprisingly long simulation time when using parallelization is not. Would you have any pointers about this ?

**Reply by mapi1** [reply]
> Another quick question: if I wanted to parallelize simulations in terms of seed (like running in parallel with same type of noise generator but different seeds), would the "correct" implementation be to add the seed as a parameter of a noise class (since noise's seed is not a params in the default classes) ? So defining for instance AdditiveNoiseParameterizedSeed(AbstractNoise) with DEFAULT_PARAMS = Bunch(sigma=0.1, seed=0), setting the jax key in the **init** function using this params.seed parameter ? Or would there be a more straightforward way ?

Regarding your question on exploring different seeds, that was not conveniently possible before. I thought about this before already and had some designs lying around which I fixed up and just released as version 0.2.9.
There is now a section in the basic tutorial that explains exactly this, running multiple seeds / keys, [see here](https://virtual-twin.github.io/tvboptim/basics/axes_and_spaces.html#scanning-noise-seeds-for-stochastic-simulations)
